File: Dopamine.SettingsModule/ViewModels/SettingsCollectionViewModel.vb
Imports Microsoft.Practices.Prism.Mvvm
Imports Dopamine.Core.Settings
Imports Microsoft.Practices.Prism
Imports Microsoft.Practices.Prism.Regions
Imports Dopamine.Common.Services.Indexing
Imports Microsoft.Practices.Prism.Commands
Imports Dopamine.Common.Services.Collection

Namespace ViewModels
    Public Class SettingsCollectionViewModel
        Inherits BindableBase
        Implements IActiveAware
        Implements INavigationAware

#Region "Variables"
        Private mIsActive As Boolean
        Private mCheckIgnoreRemovedFilesChecked As Boolean
        Private mIndexingService As IIndexingService
        Private mcollectionservice As ICollectionService
#End Region

#Region "Commands"
        Public Property RefreshNowCommand As DelegateCommand
#End Region

#Region "Properties"
        Public Property IsActive() As Boolean Implements IActiveAware.IsActive
            Get
                Return mIsActive
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mIsActive, value)
            End Set
        End Property

        Public Property CheckIgnoreRemovedFilesChecked() As Boolean
            Get
                Return mCheckIgnoreRemovedFilesChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Indexing", "IgnoreRemovedFiles", value)
                SetProperty(Of Boolean)(Me.mCheckIgnoreRemovedFilesChecked, value)
            End Set
        End Property
#End Region

#Region "Construction"
        Public Sub New(iIndexingService As IIndexingService, iCollectionService As ICollectionService)
            mIndexingService = iIndexingService
            mcollectionservice = iCollectionService

            Me.RefreshNowCommand = New DelegateCommand(AddressOf Me.RefreshNow)

            Me.GetCheckBoxesAsync()
        End Sub
#End Region

#Region "Events"
        Public Event IsActiveChanged(sender As Object, e As EventArgs) Implements IActiveAware.IsActiveChanged
#End Region

#Region "Private"
        Private Async Sub GetCheckBoxesAsync()

            Await Task.Run(Sub()
                               Me.CheckIgnoreRemovedFilesChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Indexing", "IgnoreRemovedFiles")
                           End Sub)
        End Sub

        Private Sub RefreshNow()
            mIndexingService.NeedsIndexing = True
            mIndexingService.IndexCollectionAsync(XmlSettingsClient.Instance.Get(Of Boolean)("Indexing", "IgnoreRemovedFiles"), False)
        End Sub
#End Region

#Region "INavigationAware"
        Public Function IsNavigationTarget(navigationContext As NavigationContext) As Boolean Implements INavigationAware.IsNavigationTarget
            Return True
        End Function

        Public Sub OnNavigatedFrom(navigationContext As NavigationContext) Implements INavigationAware.OnNavigatedFrom
            mIndexingService.IndexCollectionAsync(XmlSettingsClient.Instance.Get(Of Boolean)("Indexing", "IgnoreRemovedFiles"), False)
            mcollectionservice.SaveMarkedFoldersAsync()
        End Sub

        Public Sub OnNavigatedTo(navigationContext As NavigationContext) Implements INavigationAware.OnNavigatedTo

        End Sub
#End Region
    End Class
End Namespace

File: Dopamine.SettingsModule/ViewModels/SettingsCollectionViewModel.cs
using Digimezzo.Utilities.Settings;
using Dopamine.Common.Database.Repositories.Interfaces;
using Dopamine.Common.Services.Collection;
using Dopamine.Common.Services.Indexing;
using Dopamine.Common.Settings;
using Prism;
using Prism.Commands;
using Prism.Mvvm;
using Prism.Regions;
using System;
using System.Threading.Tasks;

namespace Dopamine.SettingsModule.ViewModels
{
    public class SettingsCollectionViewModel : BindableBase, IActiveAware, INavigationAware
    {
        #region Variables
        private bool isActive;
        private bool checkBoxIgnoreRemovedFilesChecked;
        private bool checkBoxRefreshCollectionAutomaticallyChecked;
        private IIndexingService indexingService;
        private ICollectionService collectionService;
        private ITrackRepository trackRepository;
        private ISettings settings;
        #endregion

        #region Commands
        public DelegateCommand RefreshNowCommand { get; set; }
        #endregion

        #region Properties
        public bool IsActive
        {
            get { return this.isActive; }
            set { SetProperty<bool>(ref this.isActive, value); }
        }

        public bool CheckBoxIgnoreRemovedFilesChecked
        {
            get { return this.checkBoxIgnoreRemovedFilesChecked; }
            set
            {
                SettingsClient.Set<bool>("Indexing", "IgnoreRemovedFiles", value);
                SetProperty<bool>(ref this.checkBoxIgnoreRemovedFilesChecked, value);
                if (!value) this.trackRepository.ClearRemovedTrackAsync(); // Fire and forget.
            }
        }

        public bool CheckBoxRefreshCollectionAutomaticallyChecked
        {
            get { return this.checkBoxRefreshCollectionAutomaticallyChecked; }
            set
            {
                this.settings.RefreshCollectionAutomatically = value;
                SetProperty<bool>(ref this.checkBoxRefreshCollectionAutomaticallyChecked, value);
            }
        }
        #endregion

        #region Construction
        public SettingsCollectionViewModel(IIndexingService indexingService, ICollectionService collectionService, 
            ITrackRepository trackRepository, ISettings settings)
        {
            this.settings = settings;
            this.indexingService = indexingService;
            this.collectionService = collectionService;
            this.trackRepository = trackRepository;

            this.RefreshNowCommand = new DelegateCommand(this.RefreshNow);

            this.GetCheckBoxesAsync();
        }
        #endregion

        #region IActiveAware
        public event EventHandler IsActiveChanged;
        #endregion

        #region Private
        private async void GetCheckBoxesAsync()
        {
            await Task.Run(() =>
            {
                this.checkBoxRefreshCollectionAutomaticallyChecked = this.settings.RefreshCollectionAutomatically;

                // Set the backing field of the property. This avoids executing a clear
                // of removed tracks when loading the screen when the setting is false.
                this.checkBoxIgnoreRemovedFilesChecked = SettingsClient.Get<bool>("Indexing", "IgnoreRemovedFiles");
                OnPropertyChanged(() => this.CheckBoxIgnoreRemovedFilesChecked);
            });
        }

        private void RefreshNow()
        {
            this.indexingService.IndexCollectionAsync();
        }
        #endregion

        #region INavigationAware
        public bool IsNavigationTarget(NavigationContext navigationContext)
        {
            return true;
        }

        public void OnNavigatedFrom(NavigationContext navigationContext)
        {
            this.indexingService.CheckCollectionAsync();
            this.collectionService.SaveMarkedFoldersAsync();
        }

        public void OnNavigatedTo(NavigationContext navigationContext)
        {
        }
        #endregion

    }
}
